ssl/tls13_enc.c

```diff
@@ -690,3 +701,58 @@ int tls13_export_keying_material(SSL *s, unsigned char *out, size_t olen,
     EVP_MD_CTX_free(ctx);
     return ret;
 }
+
+int tls13_export_keying_material_early(SSL *s, unsigned char *out, size_t olen,
+                                       const char *label, size_t llen,
+                                       const unsigned char *context,
+                                       size_t contextlen, int use_context)
+{
+    unsigned char exportsecret[EVP_MAX_MD_SIZE];
+    static const unsigned char exporterlabel[] = "exporter";
+    unsigned char hash[EVP_MAX_MD_SIZE], data[EVP_MAX_MD_SIZE];
+    const EVP_MD *md;
+    EVP_MD_CTX *ctx = EVP_MD_CTX_new();
+    unsigned int hashsize, datalen;
+    int ret = 0;
+    const SSL_CIPHER *sslcipher;
+
+    if (ctx == NULL || !ossl_statem_export_early_allowed(s))
+        goto err;
+
+    sslcipher = SSL_SESSION_get0_cipher(s->session);
+    md = ssl_md(sslcipher->algorithm2);
+
+    if (!use_context)
+        contextlen = 0;
+
+    /* In the following code, it calculates hash value of empty
+       string, and stores it in data.  This is because the definition
+       of TLS-Exporter is defined like so:
+
+       TLS-Exporter(label, context_value, key_length) =
+           HKDF-Expand-Label(Derive-Secret(Secret, label, ""),
+                             "exporter", Hash(context_value), key_length)
+
+       Derive-Secret(Secret, Label, Messages) =
+             HKDF-Expand-Label(Secret, Label,
+                               Transcript-Hash(Messages), Hash.length)
+
+       Here Transcript-Hash is the cipher suite hash algorithm. */
```

Comment:
Style: Multi-line comments should be like this
````
/*
 * This is a multi-line comment. Note that it
 * begins with a blank line.
 */

Reply:
Fixed.